[RFC] Code Agent in CodeTrans (#331)
opea-project/docs: community/rfcs/25-03-14-GenAIExample-001-CodeTrans-with-Agents.md

| Code Execution Security | Use Docker sandbox to prevent malicious code. |
| LLM Infinite Loop  | Set a max retry limit to avoid endless error regeneration. |
| Performance Issues | Parallelize Agent tasks to improve throughput. |
| User Experience    | Provide debug suggestions to reduce correction effort. |

## Implementation Plan

### Phase 1: Develop Code Execution Tool, target v1.3
- Develop the Code Execution Tool in Agent to provide a secure execution environment.
- Ensure the code execution environment is securely isolated.

### Phase 2: Core Feature Development, target v1.4
- Implement the Pre-LLM Agent for input code validation.
- Optimize the CodeTrans megaservice to automate the scheduling of Agents.
- Improve UI integration by providing a code execution interface and displaying execution results.  

### Phase 3: LLM Integration, target v1.4
- Integrate the LLM MicroService with Agent.
- Implement the Post-LLM Agent for output validation and LLM feedback handling.

### Phase 4: Optimization & Expansion, target v1.4
- Set a maximum retry limit to prevent infinite LLM regeneration loops.
- Provide debugging suggestions to enhance user experience.

## Conclusion

> By integrating Agents into the CodeTrans example, we enhance translation accuracy and stability, reduce error propagation, and improve user experience. This RFC outlines the design, architecture, and implementation plan. Future optimizations will focus on error classification, execution security, and feedback mechanisms, enabling a more intelligent and reliable code translation service.  

Reviewer: How do we measure the improvement in accuracy?

Author: The accuracy metric is not currently supported in CodeTrans example. But it can be a future task to evaluate accuracy of code-to-code tasks in GenAIEval.

Lint Check Tool:

* Proceed lint check for a snippet of code, support different coding language like `pylint`, `eslint`, `cpplint` and so on.
* Since Lint is static check, it does not require a separate execution environment, so it can be called and executed directly in a python script.

Code Execution Tool:

- Provides a secure execution environment (e.g., Docker/Sandbox) to safely run code and prevent malicious execution risks.
- For reasons of complexity of implementation, only `Python` execution tool will be supported for now.

#### Lint Check Tool

Here's a table of lint tools for different coding languages:

| Coding Language | Lint Tool     | Introduction                                                                                                                                          | Reference                                               |
| --------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------- |

> | Python          | Pylint        | A tool that checks for errors in Python code, tries to enforce a coding standard and looks for bad code smells.                                       | [link](https://www.pylint.org/)                            |

Author: Thanks, since there are too many lint checkers, we can not list all of them in this table. This table only lists the core lint tool of each language as reference.

| JavaScript      | ESLint        | A tool for identifying and reporting on patterns found in ECMAScript/JavaScript code, with the goal of making code more consistent and avoiding bugs. | [link](https://eslint.org/docs/latest/use/getting-started) |
| Java            | Checkstyle    | A development tool to help programmers write Java code that adheres to a coding standard.                                                             | [link](https://checkstyle.sourceforge.io/index.html)       |
| C++             | cpplint       | A command-line tool to check C/C++ files for style issues according to[Google&#39;s C++ style guide](http://google.github.io/styleguide/cppguide.html)   | [link](https://github.com/cpplint/cpplint)                 |

> | Go              | golangci-lint | Runs linters in parallel, reuses Go build cache and caches analysis results                                                                           | [link](https://golangci-lint.run/)                         |

Author: Great! changed to vet

> | Go              | golangci-lint | Runs linters in parallel, reuses Go build cache and caches analysis results                                                                           | [link](https://golangci-lint.run/)                         |

Reviewer 2: I think `go vet` is fine for start, but I think eventually this would need to support (also) `golangci-lint` + its linting configs.  Those could be configured either in the service backend components, or user could have control over them.

That does not need to be part of this PR though.

> | Go              | golangci-lint | Runs linters in parallel, reuses Go build cache and caches analysis results                                                                           | [link](https://golangci-lint.run/)                         |

Reviewer 2: Note: all the above listed linting tools are already available e.g. from Ubuntu package repositories, with packages having those exact names (except for `go vet` which is part of Go package).

CHECKSTYLE_JAR="./checkstyle.jar"
CHECKSTYLE_CONFIG="./google_checks.xml"

if [[ ! -f "$SOURCE_FILE" ]]; then
  echo "Source file not found: $SOURCE_FILE"
  exit 1
fi

case "$LANGUAGE" in
  python)
    echo "Running pylint..."
    pylint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
    ;;

  javascript)
    echo "Running eslint..."
    eslint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
    ;;

  java)
    echo "Running checkstyle..."
    if [[ ! -f "$CHECKSTYLE_JAR" ]]; then
      echo "Missing checkstyle.jar. Please download it first."
      exit 1
    fi
    java -jar "$CHECKSTYLE_JAR" -c "$CHECKSTYLE_CONFIG" "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
    ;;

  go)
    echo "Running golangci-lint..."
    golangci-lint run "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
    ;;

  cpp)
    echo "Running cpplint..."
    cpplint "$SOURCE_FILE" > "$REPORT_FILE" 2>&1
    ;;

  *)
    echo "Unsupported language: $LANGUAGE"
    echo "Supported language: python, javascript, java, go, cpp"
    exit 1
    ;;
esac

echo "Lint check completed. Report saved to $REPORT_FILE"

```

#### Code Execution Tool

Currently we only design to support code execution tool for `Python`.

* Prevent code injection

>   * Use Python Abstract Syntax Tree (AST) to detect and block dangerous operations such as `import os`, `exec`, and `__import__`.

Author: Providing test code for the generated code is a great idea.
But I didn't understand the exact meaning of `error he got from interacting with the program on his end`. You mean to support the functionality to fix error code from user side? or the agent detect the error code, then show it in the web UI and let users to fix it by themself?

>   * Use Python Abstract Syntax Tree (AST) to detect and block dangerous operations such as `import os`, `exec`, and `__import__`.

Reviewer 2: > You mean to support the functionality to fix error code from user side?

Yes. There are many reasons why/how user running the code locally can trigger errors that were not triggered by the CodeTrans service (regardless of whether code execution was enabled/used, or not).

It's easier for the user if s/he can just copy-paste the error to some box in the UI, and click on "Fix" button, and service takes care of re-formulating that as request for LLM and providing the (earlier) context.

Btw. While this is more relevant for CodeGen case, it's somewhat relevant also for CodeTrans: if user had modified[1] the code provided by the translation service, a way for uploading the modified target code too would also be needed.

[1] There are several reasons why user would have modified the code:
* User already fixed a bug that service missed, or continued its development on his/her own
* LLMs have limits on how large context they can handle, so user may need to provide just a snippet of that code to LLM for translation, and integrate it back to the larger system (= modify it) to properly try it out

PS. For CodeGen case, best would be if such functionality would be integrated to an IDE, so user would not need to do any copy-pasting, just click "Try running" & "Try fixing" buttons in the IDE GUI when running it locally failed, and IDE would provide all the necessary context to the service.

>   * Use Python Abstract Syntax Tree (AST) to detect and block dangerous operations such as `import os`, `exec`, and `__import__`.

Reviewer 2: > PS. For CodeGen case, best would be if such functionality would be integrated to an IDE

Meaning that APIs for the service should be clearly documented **and** version controlled because they have external users other than OPEA UI. I.e. before APIs or their semantics are changed, there's at least one OPEA release containing both old and new APIs, with old ones documented as deprecated.

>   * Use Python Abstract Syntax Tree (AST) to detect and block dangerous operations such as `import os`, `exec`, and `__import__`.

Author: @eero-t Thanks so much for the detailed descriptions!
Agree with your idea, but still think that is out of the scope of CodeTrans and should be a refined design of a complete `Code Agent`, not related to any genaiexamples.
Since this is a design of using agent in  `CodeTrans`, it's better to keep the current features. It would be greate if you'd like to create another RFC for a complete design of Code Agent which could be leveraged by all of the code-related examples.
Thanks again!

    ```python
    import ast

    code = "xxx"
    tree = ast.parse(code)
    # analyze each node in ast tree
    for node in ast.walk(tree):
        # Do import checks
    ```
* Install dependencies automatically

  * Code execution tool need to support extracting and installing dependencies from source code automatically.
  * Use AST here to extract the `import`/`from xxx import xxx` libraries.

    ```python
    import ast

    def extract_imports(code_str):
        tree = ast.parse(code_str)
        imports = set()
        for node in ast.walk(tree):
            if isinstance(node, ast.Import):
                for alias in node.names:
                    imports.add(alias.name.split('.')[0])
            elif isinstance(node, ast.ImportFrom) and node.module:
                imports.add(node.module.split('.')[0])
        return list(imports)
    ```
* Sandbox execution

  * To ensure that the code runs in a fully isolated environment, the tool needs to use container-based sandbox like `Docker`.
  * In the context of resource constraints, the resource limits (including memory, CPU, process numbers) and security policies are needed.

>   * Since the dependencies need to be installed first, the network authority will be processed in two stages:

Reviewer 2: Is the intention that the installed dependencies would be be shared between invocations?  And if yes, only for the same user, or also between different users?

Note: I'm asking because Python requirements may include conflicting versions of the dependencies, especially if they're pulled by different modules.

Author: No. the installed dependencies only work for one process. (ie. each execution of code have seperate dependencies)
The reason why I make `install requirements` a independent step is to make sure each docker container does the least amount of work. This `/code/.deps` folder will be cleaned after each execution.

    1. preperation: install the dependencies into a mounted path
       ```bash
       # add memory, cpu, process limit, and disable privilege operations
       docker run --rm -m 512m --cpus="1" --pids-limit=128 \

>          --security-opt no-new-privileges -v $(pwd)/code:/code \

Reviewer 2: This should use only the necessary capabilities, and drop all others:
`--cap-drop ALL --cap-add <what is needed>`

         sandbox-python:3.10 \
         bash -c "pip install -r /code/requirements.txt -t /code/.deps"
       ```
    2. execution: run codes using mounted dependencies
       ```bash

>        docker run --rm -v $(pwd):/code --network=none \

Reviewer 2: Why running is not secured, like dep install is?

Author: command added.

         sandbox-python:3.10 \
         python3 -I -E -S /code/user_code.py
       ```

## Expected Benefits

| Feature                      | Benefits                                              |
| ---------------------------- | ----------------------------------------------------- |
| Input Code Validation        | Catches errors early, preventing faulty translations. |
| Output Code Validation       | Ensures reliable and accurate code conversion.        |
| Automated Debug Feedback     | Reduces trial-and-error, improving LLM accuracy.      |
| Lint Static Code Check       | Catch bugs early and enforce consistent code quality. |

> | Secure Execution Environment | Protects the system from malicious code.              |

Reviewer 2: This is problem added by code execution (and potentially linting / dep install), **not** a benefit.

I think there should be a potential problem section.  For example something like below.

----

## Risks and their mitigations / user workarounds

* Node / cluster take over by execution of malicious code
  *  Mitigation: automated vetting of the executed code + its strict sandboxing
* Code execution exhausting node resources
  * Mitigation: strict resource usage limits
* Application response taking too long due to dependency install / code execution
  * Mitigation: dependency caching + enforced execution timeouts + error response to user
  * Workaround: user disables linting / code execution
* Users can affect each others' results
  * Mitigation: (dependency) caching is per-user session
* Code execution failing translation due to limits / sandboxing / dependency being offline
  * Workaround: user disables code execution / linting

| Error Classification         | Identifies syntax, logic errors for better debugging. |

## Implementation Plan

### Phase 1: Develop Code Execution Tool, target v1.3

- Develop the Code Execution Tool in Agent to provide a secure execution environment.

> - Ensure the code execution environment is securely isolated.

Reviewer 2: IMHO this is less important that the linting features, intended to cover just one of the languages, and more likely to be service security and performance issue, so it should be done later.

How the CodeTrans Helps:

- User selects both `Lint Check` and `Code Execution` in the web UI.
- Post-LLM Agent does the lint check for the translation Python code.
  - Agent will automatically fix any style/semantic issues.
- Post-LLM Agent executes the translated Python code:
  - ✅ If the code runs successfully, the system returns the output to the user.
  - ❌ If the code fails, the agent captures the error details and sends them back to the LLM.
- The LLM then retries code generation, using the error context to produce a corrected version.

This automated validation ensures that developers receive functional translations without having to manually test and debug every output.

### Preventing Infinite Regeneration Loops

Scenario:

In some cases, the LLM may repeatedly generate faulty code, leading to an endless loop of failed executions and retries. Without a safeguard, this could waste computation resources and frustrate users.

How the CodeTrans Helps:

> - Post-LLM Agent tracks retry attempts.

Reviewer 2: Linting is done in Pre step and also uses LLM, so I guess it needs this too?
```suggestion
- Both Pre- and Post-LLM Agents tracks retry attempts.
```

    Agent2 -->|Validate genereated code| CodeExecutionTool
    Agent2 -->|Output validated code| Output
```

### Components and Functionality

#### User Interface

UI Server:

- Handles user input (code, source language, target language)
- Sends requests to the CodeTrans megaservice

UI Components:

- Lint Check Button: Select to do lint check for input/output codes.
- Code Execution Button: Select to execute code for functionality check. (Support Python only)
- Input/output case: if `Code Execution Button` is selected, use will need to provide a set of input/output for this piece of code.
- Code Box – Displays the user-provided and the LLM-generated code.
- Code Translation Result – Shows the translated code, the lint check and execution result if have.

#### Backend Servers

CodeTrans Megaservice:

* Manages the scheduling of Agents, LLM, and user input/output.

Pre-LLM Agent:

- Validates code correctness, structures input/output, executes the code, and evaluates the result.
- Do lint check statically.
  - If semantic errors are detected, LLM will fix it according to lint check report.
- Runs the user-provided code to check for syntax or logical errors.

> - If errors are detected, the agent attempts to automatically fix them.

Reviewer 2: ```suggestion
- If errors are detected, the agent attempts to automatically fix them (within configurable number of attempts).
```

  - If successfully corrected, the modified code proceeds to the LLM.
  - If the errors cannot be resolved, the agent returns an error message, prompting the user to review and manually fix the code before proceeding.

LLM Microservice:

- Uses a large language model (LLM) to translate the input code into the target language.

Post-LLM Agent:

- Check code statically, executes the LLM-generated code, and verifies its correctness.
  - If execution is successful, the translated code is returned to the user.

>   - If execution fails, the error details are sent back to the LLM for regeneration (up to 3 attempts).

Reviewer 2: ```suggestion
  - If execution fails, the error details are sent back to the LLM for regeneration (within configurable number of attempts).
```
